LOOP-1057: Support for in-app and user notification alerts

rickpasetto opened the pull request:
This is a first draft of building out a user alert facility that DeviceManagers can use to post alerts either via UserNotification, via a in-app Modal, or both.  Please note that this is just the first step, there's more work to be done here (on the design side as well), but I think it is better to show this first step so that subsequent steps are smaller PRs.

This is just the LoopKit part, which introduces:
- `UserAlert`, which is a new generalized struct that encapsulates the notion of an alert, it's content, trigger, and [call-to-]actions.  The only action supported for now is "acknowledgement", but hopefully you can see from the PR that I will build on this for more actions.
- `UserAlertHandler`, a generalized protocol for, handling alert requests from devices.  I went back on forth on how to plumb this, and wound up doing it with a protocol that I've added as a sidecar to other [Delegate] protocols so plumbing it all the way from `DeviceManager`s, to their `DeviceManagerDelegate`s, through the `DeviceDataManager` (in `Loop`),  and to the (new) `UserAlertManager`.  The latter two changes are in [the PR in the _Loop_ workspace](https://github.com/tidepool-org/Loop/pull/64).
- `UserAlertResponder`, which is the reverse direction of the `Handler`.  This is what takes the response to any Action (in this case, all there is is acknowledgement), and passes it back down to the device that wants it.

I wanted to make sure there was agreement on the general approach, and then I'd hammer on the unit tests.

rickpasetto committed:
checkpoint

rickpasetto committed:
checkpoint - kind of working

rickpasetto committed:
checkpoint

rickpasetto committed:
UserAlert facility

rickpasetto committed:
tweak project file

rickpasetto committed:
tweak project file

rickpasetto committed:
checkpoint

rickpasetto committed:
tweak

rickpasetto committed:
more comments/tweaks

rickpasetto committed:
tweak

rickpasetto commented on LoopKit/DeviceManager/DeviceManager.swift:
```diff
@@ -8,20 +8,25 @@
-public protocol DeviceManagerDelegate {
+public protocol DeviceManagerDelegate: UserAlertHandler {
+    
+    #if !USE_NEW_ALERT_FACILITY
```
To make it easier to merge this, I've left the old API, and its usage, in place with this `USE_NEW_ALERT_FACILITY` marker for easy finding

ps2 replied:
Would prefer to keep unused/testing code out of dev, but if it's really too hard to change everything at once, it's ok as long as there is a concrete and short timeline for removal.

rickpasetto replied:
Yes, that's the plan.  I want to get this into `dev`, please, then post a PR for `DashKit` and Dexcom, then merge those, and then remove this code.  I am hoping by the end of the week...

rickpasetto commented on LoopKit/Notification/LoopNotificationUserInfoKey.swift:
```diff
@@ -9,5 +9,6 @@
     case bolusAmount
     case bolusStartDate
-    case cgmAlertID
+    case alertID
```
Note: I've started wiring up the ability to acknowledge from UserNotifications, but Design has not finalized for this yet.  So for now it is disabled, and all you can do is open the app.

rickpasetto committed:
minor tweak

nhamming commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -7,6 +7,8 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		1DA649AB2445174400F61E75 /* UserAlert.swift in Sources */ = {isa = PBXBuildFile; fileRef = 1DA649AA2445174400F61E75 /* UserAlert.swift */; };
+		1DD3DEB624451C3300BD8E40 /* UserAlert.swift in Sources */ = {isa = PBXBuildFile; fileRef = 1DA649AA2445174400F61E75 /* UserAlert.swift */; };
```
[nit] should there be 2?

rickpasetto replied:
Good eye.  I noticed this too.  Turns out, yes: one is a reference for the `LoopKit` target and`LoopKit-watchOS` target.  So, both are needed here.

nhamming commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,101 @@
+/// Structure that represents an Alert that needs to be shown to the User.
+public struct UserAlert {
+    /// Representation of an alert Trigger
+    public enum Trigger {
+        /// Trigger the alert immediately
+        case immediate
+        /// Delay triggering the alert by `interval`, but issue it only once.
+        case delayed(interval: TimeInterval)
+        /// Delay triggering the alert by `repeatInterval`, and repeat at that interval until cancelled or unscheduled.
+        case repeating(repeatInterval: TimeInterval)
+    }
```
Does this represent all current use cases (e.g., an alert that should be shown immediately but also repeats)?

rickpasetto replied:
There isn't a current case for an alert that is shown immediately then repeats, because right now we're using `UNTimeIntervalNotificationTrigger`, and that's how that works.  I've mirrored that functionality.  If we want to change it, let's do it in a separate PR, but I don't know of a need for it.

nhamming commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,101 @@
+        // TODO: when we have more complicated actions.  For now, all we have is "acknowledge".
+//        let actions: [UserAlertAction]
+        public let acknowledgeAction: String
+        public init(title: String, body: String, acknowledgeAction: String, isCritical: Bool = false) {
```
[nit] `acknowledgeAction` reads like this is the action to take when the alert is acknowledged. Consider updating to something like `acknowledgeButtonLabel`, which I think is closer to the intent?

rickpasetto replied:
Would you be ok with `acknowledgeActionButtonLabel`?

nhamming commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,101 @@
+
+    /// Unique device manager identifier from whence the alert came, and to which alert acknowledgements should be directed.
+    public let managerIdentifier: String
+    /// Per-alert unique identifier, for instance to group alert types.  This is the identifier that will be used to acknowledge the alert.
+    public let typeIdentifier: TypeIdentifier
```
[nit] Curious how this can be defined as unique if the type is just `String`? 

Also, does `Per-alert` mean this is a unique identifier for each alert (instead of say unique to the alert type)? If yes, adding `type` is confusing at least to me as it reads like a classification of an alert. Also, if yes, maybe `UUID` is a better type.

rickpasetto replied:
The intent is that there is a "unique identifier for each alert _type_".  This is why I created a `typealias` ... to introduce a measure of that intent.  The problem is that different devices have different 'type' identifications (Dexcom uses `Int`, DashKit uses `String`), and it is convenient to use `String` as a common form.   UUID would be very cumbersome to use for this reason, and IMHO its "uniqueness" isn't necessary for this.

nhamming replied:
Then please update the comment to reflect what you just explained. Currently it read like the opposite. For example, instead of `Per-alert unique identifier, for instance to group alert types.` maybe `Per-alertType identifier for alert grouping`. At a minimum, I think the `Per-alert` need to be updated and `unique` needs to be removed, based on your explanation.

nhamming reviewed (changes requested):
Generally looks good. I think there is some potential for confusion that could be cleaned up at this stage. For me it was the intended definition of alert or alert type.

ps2 commented on LoopKit/DeviceManager/DeviceManager.swift:
```diff
@@ -8,20 +8,25 @@
+    #endif
     func deviceManager(_ manager: DeviceManager, logEventForDeviceIdentifier deviceIdentifier: String?, type: DeviceLogEntryType, message: String, completion: ((Error?) -> Void)?)
+
+    var alertHandler: UserAlertHandler? { get set }
```
I think keeping alert handling a direct responsibility of DeviceManagerDelegate is fine, especially because it's just 3 methods, and that's how it is now.  If you want to keep it separate (maybe you want clients to be able to provide separate objects to a) handle alerts and b) handle general device delegate responsibilities), then I think it should be a separate delegate, and not nested like this.  It's also seems odd to have `DeviceManagerDelegate` conform to `UserAlertHandler` and also serve up a UserAlertHandler.

rickpasetto replied:
Yeah, it is odd, but it is the way I designed it to "pass along" the calls up the stack.  The alternative would be to pass in the `UserAlertManager` all the way down into the device managers -- which I wanted to avoid, since we already had the Delegates wired up.  Perhaps worth a quick live discussion?

And, yes, it was my intention to keep separate the "alert handling" responsibility instead of continuing to grow the responsibility of `DeviceDataManager` and the `DeviceManagerDelegate`.

ps2 replied:
As per our discussion, I do think it's fine to separate the Alert handling out of DeviceManagerDelegate. But having DeviceManagerDelegate conform to `UserAlertHandler` and also vend a `UserAlertHandler` instance seems redundant.

rickpasetto replied:
I'll take out the vending of the instance (yes, it is a little redundant).

ps2 commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,101 @@
+        /// Trigger the alert immediately
+        case immediate
+        /// Delay triggering the alert by `interval`, but issue it only once.
+        case delayed(interval: TimeInterval)
```
Do we have use cases from devices for `delayed` and `repeating`?

rickpasetto replied:
@ps2 Yes.  In Dexcom, SignalLossAlert is `repeating` and TransmitterNotFound is `delayed.`

ps2 commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,101 @@
+    /// Schedule the given alert for posting.
+    func scheduleAlert(_ alert: UserAlert)
+    /// Unschedule any pending alerts with the given identifier.
+    func unscheduleAlert(managerIdentifier: String, typeIdentifier: UserAlert.TypeIdentifier)
```
managerIdentifier does not need to be part of the protocol, it can be fetched from the manager itself, if the implementing object needs it (it may not).

ps2 replied:
I think implementing these methods on the DeviceManagerDelegate as I commented above would make this relationship more natural.

rickpasetto replied:
That's the reason I had to put it here: the implementing object doesn't need it.  I considered making a tiny `UserAlert.Identifier` struct with both of these in it to use instead but I thought it would have been overkill.   But it would have registered my intent better:  Ideally, you'd construct a `UserAlert.Identifier` at the source (with both `managerIdentifier` and `typeIdentifier`), but the "destination" user could only get out an `identifier` (of type `TypeIdentifier`) because the composition is opaque.  I liked this design better, but, again, I was thinking it might be overkill.

ps2 reviewed (changes requested):
Thanks. Looks like a good start.

rickpasetto committed:
PR Feedback

rickpasetto committed:
Wrap up the `managerIdentifier` and `typeIdentifier` into a single type, `Identifier`

rickpasetto committed:
Idea from Pete to introduce `deviceManagerInstanceIdentifier`

Eventually, we think we want to EOL the static `managerIdentifier`.

rickpasetto commented:
@ps2  @nhamming I've updated the PR with some updates, including introducing `UserAlert.Identifier` like I mention above.  Please take a look.
Unless you object to the general approach here, I will be starting on unit tests.

ps2 commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,101 @@
+//
+//  UserAlert.swift
```
We will have several sources of alerts that go to the user; I'm wondering if "Device" alerts is more specific and helps scope this work.

rickpasetto replied:
Sure, I'll change the name.

ps2 commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,101 @@
+    /// Schedule the given alert for posting.
+    func scheduleAlert(_ alert: UserAlert)
+    /// Unschedule any pending alerts with the given identifier.
+    func unscheduleAlert(managerIdentifier: String, typeIdentifier: UserAlert.TypeIdentifier)
```
unscheduleAlert and cancelAlert -> removeDeliveredAlerts and removePendingAlerts ?

rickpasetto replied:
I also never liked `scheduleAlert`, so if we're renaming `unscheduleAlert`, I prefer `issueAlert`.  I'll make all 3 changes.

ps2 replied:
I think it's schedule vs issue in the ios api because it includes a trigger that actually does the issuing.

ps2 commented on LoopKit/DeviceManager/DeviceManager.swift:
```diff
@@ -53,4 +58,9 @@ public extension DeviceManager {
         return type(of: self).localizedTitle
     }
+    
+    var deviceManagerInstanceIdentifier: String {
```
Add a comment that this is *not* currently unique per instance, but eventual intention is that it is unique per instance.

ps2 commented on LoopKit/DeviceManager/DeviceManager.swift:
```diff
@@ -53,4 +58,9 @@ public extension DeviceManager {
         return type(of: self).localizedTitle
     }
+    
+    var deviceManagerInstanceIdentifier: String {
```
Let's document that this is not unique per instance currently (and this is ok because we currently can only have one instance of each device type), but the intent is provide an attribute that would eventually be unique across the app for all instances.

nhamming replied:
ah... this answers a question I had. Why not remove `Instance` until the time that it is unique across the app? Or will that be happening in the near-future? If not, it will cause confusion until it happens and it seem that this update belongs with that coming code.

ps2 replied:
There's already a lot of confusion about what managerIdentifier is intended for.  A manager identifier identifies a certain kind of device. This is an extension of the api surface to offer an *instance* identifier. Even though it's currently implemented in a way that does not provide instance identification across multiple instances of a single manager type, it is effective within Loop use as Loop can only have one of each type of device.  When we have the need to support multiple instances of a single type, the implementation can change, and the uses do not need to change.

rickpasetto replied:
Removed `Instance` and added a comment.

ps2 commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,108 @@
+        /// Unique device manager identifier from whence the alert came, and to which alert acknowledgements should be directed.
+        public let deviceManagerInstanceIdentifier: String
+        /// Per-alert unique identifier, for instance to group alert types.  This is the identifier that will be used to acknowledge the alert.
+        public let typeIdentifier: TypeIdentifier
```
alertTypeIdentifier?

rickpasetto replied:
I felt that it was kind of redundant, since this already was in a subtype of `UserAlert`.

ps2 commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,108 @@
+//
+//  UserAlert.swift
```
We have other alerts that will be shown to the user. This work pertains to device alerts, so maybe naming this "DeviceAlert" would help describe and scope this.

ps2 commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,108 @@
+    /// Schedule the given alert for posting.
+    func scheduleAlert(_ alert: UserAlert)
+    /// Unschedule any pending alerts with the given identifier.
+    func unscheduleAlert(identifier: UserAlert.Identifier)
```
unscheduleAlert and cancelAlert -> removePendingAlertRequest and removeDeliveredAlert?

rickpasetto replied:
I changed it to `removePendingAlert` and `removeDeliveredAlert`...hope that's ok, it felt more symmetric.

ps2 commented on LoopKit/UserAlert.swift:
```diff
@@ -0,0 +1,108 @@
+    /// This type represents a per-alert-type identifier, but not necessarily unique across devices.  Each device may have its own Swift type for this,
+    /// so conversion to String is the most convenient, but aliasing the type is helpful because it is not just "any String".
+    public typealias TypeIdentifier = String
+    public typealias AcknowledgeCompletion = (TypeIdentifier) -> Void
```
Does this support a different use from UserAlertResponder.acknowledgeAlert() ?  Also, supporting persistence could be hard for this since it's being used below in a closure.

rickpasetto replied:
Yeah, I went back and forth on this.  I think I'll get rid of this in favor of AlertResponder.

rickpasetto committed:
PR Feedback: renaming:

UserAlert* -> DeviceAlert*
scheduleAlert -> issueAlert
unscheduleAlert -> removePendingAlerts
cancelAlert -> removeDeliveredAlerts

rickpasetto committed:
More PR feedback, and some other tweaks

rickpasetto committed:
Oops, forgot project file

rickpasetto committed:
Oops again, a little too much project file :)

rickpasetto committed:
More PR Feedback: get rid of AcknowledgeCompletion

nhamming commented on LoopKit/DeviceAlert.swift:
```diff
@@ -43,45 +43,48 @@ public struct UserAlert {
             self.isCritical = isCritical
         }
     }
+    public struct Identifier: Equatable {
+        /// Unique device manager identifier from whence the alert came, and to which alert acknowledgements should be directed.
+        public let deviceManagerInstanceIdentifier: String
+        /// Per-alert unique identifier, for instance to group alert types.  This is the identifier that will be used to acknowledge the alert.
+        public let typeIdentifier: TypeIdentifier
+        public init(deviceManagerInstanceIdentifier: String, typeIdentifier: TypeIdentifier) {
+            self.deviceManagerInstanceIdentifier = deviceManagerInstanceIdentifier
+            self.typeIdentifier = typeIdentifier
+        }
+        /// An opaque value for this tuple for unique identification of the alert across devices.
+        public var value: String {
+            return "\(deviceManagerInstanceIdentifier).\(typeIdentifier)"
+        }
+    }
```
[nit] since the struct is called `Identifier` could `typeIdentifier` be reduced to `type` and `deviceManagerInstanceIdentifier` be reduced to `deviceManager`? (note this takes into account my comment about identifying device manager instances.

rickpasetto replied:
@nhamming I discussed the name of this variable at length with @ps2 .  This isn't a `DeviceManager`, so `deviceManager` won't work.  I tried `deviceIdentifier` but that name is already taken for other things.  Since Pete and I agreed it is really the identifier for the "instance", hence the (somewhat wordy) `deviceManagerInstanceIdentifier`.  I would be fine with `deviceManagerIdentifier`, though.  WDYT?

nhamming replied:
Here my comment is the `Identifier` part is unneeded, since the struct is `Identifier` (not add on to my comment about using `Instance`). So, it would be `Identifier.deviceManager` or `Identifier.type` instead of `Identifier.deviceManagerIdentifier` or `Identifier.typeIdentifier`. But I'm not holding strong to this comment.

nhamming reviewed (approved):
1 strong suggestion related to the device manager instance identifier, but not a blocker. Otherwise LGTM.

rickpasetto committed:
PR feedback: renaming only

deviceManagerInstanceIdentifier -> deviceManagerIdentifier
removePendingAlerts -> removePendingAlert
removeDeliveredAlerts -> removeDeliveredAlert

ps2 commented on LoopKit/DeviceManager/DeviceManager.swift:
```diff
@@ -53,4 +56,9 @@ public extension DeviceManager {
         return type(of: self).localizedTitle
     }
+    
+    var deviceManagerIdentifier: String {
```
Could also just be `managerIdentifier`, as the context (class vs instance) here doesn't matter; it is really the same thing, and the same meaning.

rickpasetto replied:
heh that's what I had in the first place...I'll change it back /shrug

rickpasetto committed:
More PR Feedback

rickpasetto committed:
PR feedback: one more rename: deviceManagerIdentifier -> managerIdentifier

nhamming reviewed (approved):
LGTM
